### PR TITLE
Show up sign in form in standalone

### PR DESCRIPTION
**Pull request by un-tone**
For https://github.com/StarcounterApps/Calendar/issues/21.
Implemented with [authorization](https://github.com/Starcounter/authorization) library.

### CODE REVIEW

File: src/Calendar/Api/AuthorizedHandlers.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            router.AddMiddleware(new SecurityMiddleware(enforcement,
+                info =>
+                {
+                    MasterPage master = LauncherHooks.GetMaster();
+
+                    if (authenticationRequiredPages.Contains(info.SelectedPageType))
+                    {
+                        var returnUri = Utils.RetrieveReturnUri(info.Request.Uri);
+
+                        master.CurrentPage = new RedirectPage
```
Comment (joozek78) [inline src/Calendar/Api/AuthorizedHandlers.cs]: can you just put `returnUri = ???` into the if? Even better to use ternary operator

File: src/Calendar/Api/AuthorizedHandlers.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                return Self.GET(partialUri);
+            });
+
+            router.AddMiddleware(new MasterPageMiddleware());
+
+            var rules = RegisterRules();
+            var enforcement = new AuthorizationEnforcement(rules, new SystemUserAuthentication());
+
+            router.AddMiddleware(new SecurityMiddleware(enforcement,
+                info =>
```
Comment (joozek78) [inline src/Calendar/Api/AuthorizedHandlers.cs]: I would put that block into a method. It's a little confusing to have such a large block here

File: src/Calendar/Api/LauncherHooks.cs [hunk 3]
```diff
@@ -75,9 +56,22 @@ public class LauncherHooks {
...
-        static public MasterPage GetMaster() {
-
-            return Self.GET<MasterPage>("/Calendar/standalone");
+        public static MasterPage GetMaster()
+        {
+            Session session = Session.Current;
+
+            if (session?.Data != null)
+            {
+                return session.Data as MasterPage;
```
Comment (joozek78) [inline src/Calendar/Api/LauncherHooks.cs]: what if `session.Data` is not `MasterPage`?

File: src/Calendar/Api/Permissions/AuthenticatedAccessPermission.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿using Starcounter.Authorization.Core;
+
+namespace Calendar.Api.Permissions
+{
+    public class AuthenticatedAccessPermission : Permission
```
Comment (joozek78) [inline src/Calendar/Api/Permissions/AuthenticatedAccessPermission.cs]: This is not a good permission. Permission should describe a specific action that can be authorized

File: src/Calendar/Helpers/Utils.cs [hunk 5]
```diff
@@ -50,5 +51,21 @@ public class Utils {
         //        public static CanUserAdmin( RoomBooking.Database.Reservation reservation,
```
Comment (joozek78) [inline src/Calendar/Helpers/Utils.cs]: can you remove this commented code while you're here?

File: src/Calendar/Helpers/Utils.cs [hunk 6]
```diff
@@ -50,5 +51,21 @@ public class Utils {
...
+            string returnUri = uri;
+            string query = HttpUtility.UrlDecode(uri);
+            string[] queryParts = query?.Split('?');
+            if (queryParts?.Length > 1)
+            {
+                var queryParams = HttpUtility.ParseQueryString(queryParts[1]);
+                var newReturnUri = queryParams.Get("return_uri");
+                if (!string.IsNullOrEmpty(newReturnUri))
+                {
+                    returnUri = newReturnUri;
```
Comment (joozek78) [inline src/Calendar/Helpers/Utils.cs]: why not inline `returnUri` and put a return statement here?

File: src/Calendar/ViewModels/PermissionDeniedPage.json.cs [hunk 7]
```diff
@@ -1,7 +1,15 @@
...
+using Starcounter.Authorization.Attributes;
+using Starcounter.Authorization.Routing.Middleware;
+using Calendar.Api.Attributes;
+using Calendar.Api.Permissions;
+
+namespace Calendar.ViewModels
+{
+    [PartialUrl("/calendar/partial/restrict?return_uri={?}")]
+    [UseDbScope(false)]
+    [RequirePermission(typeof(AuthenticatedAccessPermission))]
```
Comment (joozek78) [inline src/Calendar/ViewModels/PermissionDeniedPage.json.cs]: I don't think this page should have a restricted access...

File: src/Calendar/ViewModels/SettingsPage.json.cs [hunk 8]
```diff
@@ -1,11 +1,13 @@
...
-using System.Text;
+using Starcounter.Authorization.Attributes;
+using Calendar.Api.Attributes;
+using Calendar.Api.Permissions;
+using Calendar.Helpers;
 
 namespace Calendar.ViewModels {
-
+    [PartialUrl("/calendar/partial/settings")]
+    [RequirePermission(typeof(SettingsPermission))]
```
Comment (joozek78) [inline src/Calendar/ViewModels/SettingsPage.json.cs]: `EditSettings` would be a better permission name

File: src/Calendar/Api/AuthorizedHandlers.cs (hunk 2 above)
Comment (un-tone) [inline src/Calendar/Api/AuthorizedHandlers.cs]: Done.

File: src/Calendar/Api/AuthorizedHandlers.cs (hunk 1 above)
Comment (un-tone) [inline src/Calendar/Api/AuthorizedHandlers.cs]: Done.

File: src/Calendar/Api/LauncherHooks.cs (hunk 3 above)
Comment (un-tone) [inline src/Calendar/Api/LauncherHooks.cs]: Fixed this.

File: src/Calendar/Helpers/Utils.cs (hunk 5 above)
Comment (un-tone) [inline src/Calendar/Helpers/Utils.cs]: Done.

File: src/Calendar/Helpers/Utils.cs (hunk 6 above)
Comment (un-tone) [inline src/Calendar/Helpers/Utils.cs]: Done.

File: src/Calendar/ViewModels/SettingsPage.json.cs (hunk 8 above)
Comment (un-tone) [inline src/Calendar/ViewModels/SettingsPage.json.cs]: Fixed.

File: src/Calendar/Api/Permissions/AuthenticatedAccessPermission.cs (hunk 4 above)
Comment (un-tone) [inline src/Calendar/Api/Permissions/AuthenticatedAccessPermission.cs]: Renamed it to `RestrictAccessAbsolutely`

File: src/Calendar/ViewModels/PermissionDeniedPage.json.cs (hunk 7 above)
Comment (un-tone) [inline src/Calendar/ViewModels/PermissionDeniedPage.json.cs]: I used this to define should the app redirect to the authentication page or not.

File: src/Calendar/Api/AuthorizedHandlers.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+{
+    class AuthorizedHandlers
+    {
+        public void Register()
+        {
+            Type[] assemblyTypes = Assembly.GetAssembly(typeof(AuthorizedHandlers)).GetTypes();
+
+            var accessRestrictionPages = assemblyTypes
+                .Select(type => Tuple.Create(type, type.GetCustomAttribute<RequirePermissionAttribute>()))
+                .Where(x => x.Item2 != null && x.Item2.RequiredPermission == typeof(RestrictAccessAbsolutely))
```
Comment (joozek78) [inline src/Calendar/Api/AuthorizedHandlers.cs]: That's not a correct usage of permissions. Define a new attribute and use it.
BTW you could make it shorter: `.Where(x => x.Item2?.RequiredPermission == typeof(RestrictAccessAbsolutely))`

File: src/Calendar/Api/AuthorizedHandlers.cs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace Calendar.Api
+{
+    class AuthorizedHandlers
+    {
+        public void Register()
+        {
+            Type[] assemblyTypes = Assembly.GetAssembly(typeof(AuthorizedHandlers)).GetTypes();
+
+            var accessRestrictionPages = assemblyTypes
```
Comment (joozek78) [inline src/Calendar/Api/AuthorizedHandlers.cs]: why not move it into `CreateOutsideFacingRouter`?

File: src/Calendar/Api/Permissions/RestrictAccessAbsolutely.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿using Starcounter.Authorization.Core;
+
+namespace Calendar.Api.Permissions
+{
+    public class RestrictAccessAbsolutely : Permission
```
Comment (joozek78) [inline src/Calendar/Api/Permissions/RestrictAccessAbsolutely.cs]: That's not a good a Permission. It doesn't describe an action

File: src/Calendar/Api/Permissions/AuthenticatedAccessPermission.cs (hunk 4 above)
Comment (joozek78) [inline src/Calendar/Api/Permissions/AuthenticatedAccessPermission.cs]: If you want to have one permission for all of the "basic" actions I suggest `OpenBasicPages`. Permission should describe the action that we can control, not the act of restriction. It's best imagined that it's a label of a button - "List Images" button makes sense, but "Restrict Access Absolutely" not so much

File: src/Calendar/Helpers/SessionHelper.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Calendar.Helpers
+{
+    public class SessionHelper
+    {
+        public static MasterPage GetMaster()
+        {
+            Session session = Session.Current ?? new Session(SessionOptions.PatchVersioning);
+
+            var master = session?.Data as MasterPage;
```
Comment (joozek78) [inline src/Calendar/Helpers/SessionHelper.cs]: `return (session?.Data as MasterPage) ?? new MasterPage {Session = session}`

File: src/Calendar/Helpers/Utils.cs [hunk 13]
```diff
@@ -1,54 +1,64 @@
...
-        public static Simplified.Ring5.SystemUserSession GetCurrentUserSession() {
-
-            if (Session.Current == null) {
+using Calendar.Api;
+
+namespace Calendar.Helpers
+{
+    public class Utils
+    {
+        public static SystemUserSession GetCurrentUserSession()
```
Comment (joozek78) [inline src/Calendar/Helpers/Utils.cs]: I think there's a method for that in simplified... https://github.com/starcounterprefabs/simplified/blob/988bfb7e79ae569ce949359140d42a96575b25f4/Ring3/User/SystemUser.Static.cs#L22

File: src/Calendar/ViewModels/PermissionDeniedPage.json.cs (hunk 7 above)
Comment (joozek78) [inline src/Calendar/ViewModels/PermissionDeniedPage.json.cs]: But still... I think it's confusing to restrict access to this page. Not a big deal

File: src/Calendar/Api/AuthorizedHandlers.cs (hunk 9 above)
Comment (joozek78) [inline src/Calendar/Api/AuthorizedHandlers.cs]: BTW2 I think the we should select "Please log in" or "You are not allowed" page based on whether  the current user is logged in, not on what kind of page he is trying to access, WDYT?

File: src/Calendar/ViewModels/PermissionDeniedPage.json.cs (hunk 7 above)
Comment (joozek78) [inline src/Calendar/ViewModels/PermissionDeniedPage.json.cs]: But still... I think it's confusing to restrict access to this page. Not a big deal

File: src/Calendar/Api/AuthorizedHandlers.cs (hunk 10 above)
Comment (un-tone) [inline src/Calendar/Api/AuthorizedHandlers.cs]: Got rid of this code.

File: src/Calendar/Api/AuthorizedHandlers.cs (hunk 9 above)
Comment (un-tone) [inline src/Calendar/Api/AuthorizedHandlers.cs]: Yes, I implemented so. Please check. Currently I see that my initial conception was pretty wrong

File: src/Calendar/Api/Permissions/RestrictAccessAbsolutely.cs (hunk 11 above)
Comment (un-tone) [inline src/Calendar/Api/Permissions/RestrictAccessAbsolutely.cs]: OK

File: src/Calendar/Helpers/SessionHelper.cs (hunk 12 above)
Comment (un-tone) [inline src/Calendar/Helpers/SessionHelper.cs]: Fixed.

File: src/Calendar/Helpers/Utils.cs (hunk 13 above)
Comment (un-tone) [inline src/Calendar/Helpers/Utils.cs]: Removed this code

File: src/Calendar/ViewModels/PermissionDeniedPage.json.cs (hunk 7 above)
Comment (warpech) [inline src/Calendar/ViewModels/PermissionDeniedPage.json.cs]: Yeah it looks weird. We require an authenticated user to see the authentication page?

**Comment by un-tone** [comment]
The PR is ready for review

File: src/Calendar/ViewModels/SettingsPage.json.cs [hunk 14]
```diff
@@ -5,10 +5,12 @@
 using Calendar.Api.Attributes;
 using Calendar.Api.Permissions;
 using Calendar.Helpers;
+using Starcounter.Authorization.Routing.Middleware;
 
 namespace Calendar.ViewModels
 {
     [PartialUrl("/Calendar/partial/settings")]
+    [UseDbScope]
```
Comment (un-tone) [inline src/Calendar/ViewModels/SettingsPage.json.cs]: Removed.

**Comment by joozek78** [comment]
I double-approve it :)
The reason I don't merge PRs for anyone is that it's ultimately the author who needs to be sure about it, you could've some changes in mind that I don't know about

**Pull request merged**